Tweak ESLint to use our custom eslint-config-universal-search config

pdehaan opened the pull request.

pdehaan commented on bootstrap.js:
```diff
@@ -14,9 +14,9 @@ function startup(data, reason) {
+  if (reason === APP_SHUTDOWN) { return; }
 
-  if (reason == ADDON_DISABLE || reason == ADDON_UNINSTALL) {
+  if (reason === ADDON_DISABLE || reason === ADDON_UNINSTALL) {
```
I have no idea what I'm doing here... Plus, ESLint complains these are undefined. If we cant use `===` here, I can work on silencing the ESLint warnings for these lines/blocks.

jaredhirsch replied:
Yeah, this is a bit weird for sure. Welcome to XUL-landia.

The `reason`s are defined by the code that runs the bootstrap file, such that the constants should be predefined when the code executes: https://developer.mozilla.org/en-US/Add-ons/Bootstrapped_extensions#Reason_constants

jaredhirsch replied:
This is probably a case where it would be nice to define the constants as ignored globals at the top of the file, I think.

pdehaan replied:
Done.

pdehaan commented on lib/ui/Popup.js:
```diff
@@ -199,13 +200,13 @@ Popup.prototype = {
           title: controller.getCommentAt(i),
           type: controller.getStyleAt(i),
-          text: controller.searchString.replace(/^\s+/, "").replace(/\s+$/, "")
+          text: controller.searchString.replace(/^\s+/, '').replace(/\s+$/, '')
```
I'm not sure what the regex is happening here. Not sure if we can just use this to strip the leading/trailing whitespaces:

``` js
text: controller.searchString.trim()
```

jaredhirsch replied:
This code comes to us from the [sick, sad world of XBL](https://github.com/mozilla/gecko-dev/blob/master/toolkit/content/widgets/autocomplete.xml#L1156-L1157), and in fact those lines were originally created before `String.prototype.trim` was a thing (added to FF in version 3.5).

So yeah, I should fix that. Or you're welcome to.

pdehaan commented on lib/ui/Popup.js:
```diff
@@ -237,7 +238,7 @@ Popup.prototype = {
     // searchTerm is the same thing as the 'text' item sent down in each result.
     // maybe that's not a useful place to put the search term...
-    let searchTerm = controller.searchString.replace(/^\s+/, "").replace(/\s+$/, "")
+    let searchTerm = controller.searchString.replace(/^\s+/, '').replace(/\s+$/, '');
```
Same as [above](https://github.com/mozilla/universal-search-addon/pull/39/files#r34214814). Can we just use `controller.searchString.trim()`?

jaredhirsch replied:
yup. /me tattoos a tear under one eye to mark the first year with XBL

jaredhirsch commented on package.json:
```diff
@@ -10,7 +10,8 @@
   "devDependencies": {
-    "eslint": "0.24.0"
+    "eslint": "0.24.0",
+    "eslint-config-universal-search": "1.0.0"
```
noice

pdehaan commented:
Anyways, kind of boring PR. Mostly just me tweaking some double->single quotes and trying to prefer `===` over `==`, and a few whitespace tweaks.

jaredhirsch commented on .eslintrc:
```diff
@@ -17,9 +15,4 @@ globals:
   XPCOMUtils: false
```
This isn't like the _most_ awesome thing ever, since we have a "global" for each module / service.

Is it possible it would be cleaner to include comments at the top of each file with the list of modules? That would tend to be close to where the modules are imported, so I might have a marginally better shot at remembering to update this over time.

I'm not sure of the answer here, curious what you think.

pdehaan replied:
I'd be hard pressed to care any less... I think mostly depends on if you want to define once in one .eslitnrc (and have them leak everywhere), or maintain the globals in separate files.

If you prefer no globals in the eslintrc file, it's a quick fix for me to disable them and then silence the errors in the specific files.

I'm here to serve.

jaredhirsch commented:
lgtm, do what you will about trim() and the globals. Feel free to merge this yourself if you're like, "meh..."

I'm happy to help you get a development environment setup for actually running this code, or maybe you've done addon dev before and the [setup steps](https://github.com/mozilla/universal-search-addon/blob/master/README.md#basic-addon-environment) don't seem like a ridiculous affront to your dignity. Let me know.

pdehaan commented:
r? @6a68 

I removed the leaky globals from .eslintrc and put them in the specific files (and switched from dual regexes to `trim()` since I saw we were using `trim()` elsewhere anyways).

pdehaan commented:
And for the sake of full transparency, here's the current output from ESLint:

```
$$ npm run lint

> mozilla-universal-search-addon@1.0.1 lint /Users/pdehaan/dev/github/universal-search-addon
> eslint .


bootstrap.js
   5:17  error  Cc is defined but never used         no-unused-vars
   5:33  error  Ci is defined but never used         no-unused-vars
   5:57  error  Cm is defined but never used         no-unused-vars
  11:9   error  startup is defined but never used    no-unused-vars
  16:9   error  shutdown is defined but never used   no-unused-vars
  32:9   error  install is defined but never used    no-unused-vars
  38:9   error  uninstall is defined but never used  no-unused-vars

lib/main.js
    7:17  error  Cc is defined but never used                no-unused-vars
    7:33  error  Ci is defined but never used                no-unused-vars
    7:57  error  Cm is defined but never used                no-unused-vars
   18:4   error  EXPORTED_SYMBOLS is defined but never used  no-unused-vars
   86:6   error  document is defined but never used          no-unused-vars
  143:4   error  Main is defined but never used              no-unused-vars

lib/ui/GoButton.js
  5:4  error  EXPORTED_SYMBOLS is defined but never used  no-unused-vars

lib/ui/Popup.js
   14:4  error  EXPORTED_SYMBOLS is defined but never used  no-unused-vars
  257:4  error  Expected no return value                    consistent-return

lib/ui/Urlbar.js
  7:4  error  EXPORTED_SYMBOLS is defined but never used  no-unused-vars

✖ 17 problems (17 errors, 0 warnings)
```

(Not saying that these need to be addressed in this PR, but it'd be nice to get ESLint passing so we can hook up Travis or something.)

jaredhirsch commented:
Let's see. We're stuck with EXPORTED_SYMBOLS everywhere, it's like `exports`. Related, 'Main' has to stay; it's the thing which is exported. Also, bootstrap.js has to have exactly those four functions with the unused parameters. (Don't ask me why, I have no idea.) 

On the other hand, Cc, Ci, Cm could go if they aren't used; I just insert them out of habit when doing gecko things. And I guess document isn't needed in main.js.

Finally, I'm not sure what the consistent-return error is in Popup.

If you want to address those, feel free. Otherwise, you're welcome to merge.

pdehaan commented:
Merging. Lets deal with it another day when it isn't 6:39pm.

jaredhirsch commented:
![](http://media.giphy.com/media/2xO491sY6f0cM/giphy.gif)

pdehaan commented:
> @6a68: Finally, I'm not sure what the `consistent-return` error is in Popup.

---

> `consistent-return` - require `return` statements to either always or never specify values
> &mdash; via http://eslint.org/docs/rules/consistent-return

Basically, ESLint is warning you that your function either returns `undefined` or maybe a Promise, and that can be risky bidness:

``` js
233:    if (!Services.search.isInitialized) {
234:      return;
235:    }
// ...
256:    // returns a promise for the formatted results of the search suggestion engine
257:    return suggestions;
258:  }
```

via [./lib/ui/Popup.js:233-258](https://github.com/mozilla/universal-search-addon/blob/d1a676da09fc205d3f74e84b080b64770cf13b29/lib/ui/Popup.js#L233-L258)

I cheated in #41 and just disabled the `consistent-return` rule globally. But we can find other solutions if you want. Or we just merge #41 and ponder it as a future z-later issue.
